File: src/DIY/State/State.java
package DIY.State;

public class State {
    public static void main(String[] args) {

        // // State DIY
        var player = new Player();

        player.play();
        player.pause();
        player.play();
        player.stop();
    }

}
class Player {

    private Action state;
    
    public Player() {
        this.state = new StoppedState();
    }
    public void setState(Action state) {
        this.state = state;
    }
    public void play(){
        state.play(this);
    }
    public void pause(){
        state.pause(this);

    }
    public void stop(){
        state.stop(this);

    }

}

interface Action {
    void play(Player player);
    void pause(Player player);
    void stop(Player player);
}

class PausedState implements Action{

    @Override
    public void play(Player player) {
        System.out.println("Resuming playback...");
        player.setState(new PlayState());
    }

    @Override
    public void pause(Player player) {
        System.out.println("Playback is already paused.");
    }

    @Override
    public void stop(Player player) {
        System.out.println("Stopping playback...");
        player.setState(new StoppedState());
        System.out.println("Playback stopped.");
    }

}
class PlayState implements Action {

    @Override
    public void play(Player player) {
        System.out.println("Already playing.");
    }

    @Override
    public void pause(Player player) {
        System.out.println("Pausing playback...");
        player.setState(new PausedState());
        
        System.out.println("Playback paused.");
    }

    @Override
    public void stop(Player player) {
        System.out.println("Stopping audio...");
        player.setState(new StoppedState());
        System.out.println("Playback stopped.");
    }

}
class StoppedState implements Action {

    @Override
    public void play(Player player) {
        System.out.println("Starting playback...");
        player.setState(new PlayState());
    }

    @Override
    public void pause(Player player) {
        System.out.println("Cannot pause playback while stopped.");
    }

    @Override
    public void stop(Player player) {
        System.out.println("Already stopped.");
    }

}

File: src/DIY/State/State.ts
// State.ts

class Player {
    private state: Action;

    constructor() {
        this.state = new StoppedState();
    }

    public setState(state: Action): void {
        this.state = state;
    }

    public play(): void {
        this.state.play(this);
    }

    public pause(): void {
        this.state.pause(this);
    }

    public stop(): void {
        this.state.stop(this);
    }
}

interface Action {
    play(player: Player): void;
    pause(player: Player): void;
    stop(player: Player): void;
}

class PausedState implements Action {
    play(player: Player): void {
        console.log("Resuming playback...");
        player.setState(new PlayState());
    }

    pause(player: Player): void {
        console.log("Playback is already paused.");
    }

    stop(player: Player): void {
        console.log("Stopping playback...");
        player.setState(new StoppedState());
        console.log("Playback stopped.");
    }
}

class PlayState implements Action {
    play(player: Player): void {
        console.log("Already playing.");
    }

    pause(player: Player): void {
        console.log("Pausing playback...");
        player.setState(new PausedState());
        console.log("Playback paused.");
    }

    stop(player: Player): void {
        console.log("Stopping audio...");
        player.setState(new StoppedState());
        console.log("Playback stopped.");
    }
}

class StoppedState implements Action {
    play(player: Player): void {
        console.log("Starting playback...");
        player.setState(new PlayState());
    }

    pause(player: Player): void {
        console.log("Cannot pause playback while stopped.");
    }

    stop(player: Player): void {
        console.log("Already stopped.");
    }
}

// main
(function main() {
    const player = new Player();

    player.play();   // Start playback
    player.pause();  // Pause playback
    player.play();   // Resume
    player.stop();   // Stop
})();
